odl/trafos/util/ft_utils.py

```diff
@@ -0,0 +1,654 @@
+    case.
+
+    Parameters
+    ----------
+    recip_grid : `RegularGrid`
+        Sampling grid in reciprocal space.
+    x0 : `array-like`
+        Desired minimum point of the real space grid.
+    axes : int or `sequence` of ints, optional
+        Dimensions in which to calculate the real space grid. The sequence
+        must have the same length as ``shift`` if the latter is given
+        as a sequence. ``None`` means "all axes".
+    halfcomplex : bool, optional
+        If ``True``, interpret the given grid as the reciprocal as used
+        in a half-complex FFT (see above). Otherwise, the grid is
+        regarded as being used in a complex-to-complex transform.
+    halfcx_parity : {'even', 'odd'}
+        Use this parity for the shape of the returned grid in the
+        last axis of ``axes`` in the case ``halfcomplex=True``
+
+    Returns
+    -------
+    irecip : `RegularGrid`
+        The inverse reciprocal grid
+    """
+    if axes is None:
+        axes = list(range(recip_grid.ndim))
+    else:
+        try:
+            axes = [int(axes)]
+        except TypeError:
+            axes = list(axes)
+
+    rstride = recip_grid.stride
+    rshape = recip_grid.shape
+
+    # Calculate shape of the output grid by adjusting in axes[-1]
+    irshape = list(rshape)
+    if halfcomplex:
+        if str(halfcx_parity).lower() == 'even':
+            irshape[axes[-1]] = 2 * rshape[axes[-1]] - 2
+        elif str(halfcx_parity).lower() == 'odd':
+            irshape[axes[-1]] = 2 * rshape[axes[-1]] - 1
+        else:
+            raise ValueError("`halfcomplex` parity '{}' not understood"
+                             "".format(halfcx_parity))
+
+    irmin = np.asarray(x0)
+    irshape = np.asarray(irshape)
+    irstride = np.copy(rstride)
+    irstride[axes] = 2 * np.pi / (irshape[axes] * rstride[axes])
+    irmax = irmin + (irshape - 1) * irstride
+
+    return RegularGrid(irmin, irmax, irshape)
+
+
+def dft_preprocess_data(arr, shift=True, axes=None, sign='-', out=None):
+    """Pre-process the real-space data before DFT.
+
+    This function multiplies the given data with the separable
+    function::
+
+        p(x) = exp(+- 1j * dot(x - x[0], xi[0]))
+
+    where ``x[0]`` and ``xi[0]`` are the minimum coodinates of
+    the real-space and reciprocal grids, respectively. The sign of
+    the exponent depends on the choice of ``sign``. In discretized
+    form, this function becomes an array::
+
+        p[k] = exp(+- 1j * k * s * xi[0])
+
+    If the reciprocal grid is not shifted, i.e. symmetric around 0,
+    it is ``xi[0] =  pi/s * (-1 + 1/N)``, hence::
+
+        p[k] = exp(-+ 1j * pi * k * (1 - 1/N))
+
+    For a shifted grid, we have :math:``xi[0] =  -pi/s``, thus the
+    array is given by::
+
+        p[k] = (-1)**k
+
+    Parameters
+    ----------
+    arr : `array-like`
+        Array to be pre-processed. If its data type is a real
+        non-floating type, it is converted to 'float64'.
```

Comment:
No, I don't convert in that case. It only applies for stuff like `int` or `bool`.